File: app/src/main/java/com/example/ugcssample/utils/ArrayUtils.java
package com.example.ugcssample.utils;

import java.util.ArrayList;
import java.util.List;

public final class ArrayUtils {

    private ArrayUtils() {
    }

    public static boolean isEmpty(List<?> list) {
        return list == null || list.isEmpty();
    }

    public static <T> List<T> copy(List<T> src) {
        if (src == null)
            return null;
        List<T> retVal = new ArrayList<T>(src.size());
        for (T element : src) {
            retVal.add(element);
        }
        return retVal;
    }

    public static double[] copy(double[] src) {
        if (src == null)
            return null;
        double[] retVal = new double[src.length];
        for (int i = 0; i < src.length; i++) {
            retVal[i] = src[i];
        }
        return retVal;
    }

    public static double[] newDouble(int size, double val) {
        double[] retVal = new double[size];
        for (int i = 0; i < size; i++) {
            retVal[i] = val;
        }
        return retVal;
    }

    public static boolean[] newBoolean(int size, boolean val) {
        boolean[] retVal = new boolean[size];
        for (int i = 0; i < size; i++) {
            retVal[i] = val;
        }
        return retVal;
    }

    public static void setBoolean(boolean[] retVal, boolean val) {
        for (int i = 0; i < retVal.length; i++) {
            retVal[i] = val;
        }
    }

    public static Double findMax(double[] values) {
        Double d = null;
        if (values != null) {
            for (double v : values) {
                if (d == null || d < v)
                    d = v;
            }
        }
        return d;
    }

    public static Double findMin(double[] values) {
        Double d = null;
        if (values != null) {
            for (double v : values) {
                if (d == null || d > v)
                    d = v;
            }
        }
        return d;
    }

    public static List<Double> clone(ArrayList<Double> src) {
        if (src == null) {
            return null;
        }
        return new ArrayList<>(src);
    }

    public static void copy(double[] origin, double[] target, double valueToAdd) {
        for (int i = 0; i < origin.length; i++) {
            target[i] = origin[i] + valueToAdd;
        }
    }

}

File: app/src/main/java/com/example/ugcssample/utils/ArrayUtils.kt
package com.example.ugcssample.utils

import java.util.*

object ArrayUtils {
    fun isEmpty(list: MutableList<*>?): Boolean {
        return list == null || list.isEmpty()
    }

    fun <T> copy(src: MutableList<T?>?): MutableList<T?>? {
        if (src == null) return null
        val retVal: MutableList<T?> = ArrayList(src.size)
        for (element in src) {
            retVal.add(element)
        }
        return retVal
    }

    fun copy(src: DoubleArray?): DoubleArray? {
        if (src == null) return null
        val retVal = DoubleArray(src.size)
        for (i in src.indices) {
            retVal[i] = src[i]
        }
        return retVal
    }

    fun newDouble(size: Int, `val`: Double): DoubleArray? {
        val retVal = DoubleArray(size)
        for (i in 0 until size) {
            retVal[i] = `val`
        }
        return retVal
    }

    fun newBoolean(size: Int, `val`: Boolean): BooleanArray? {
        val retVal = BooleanArray(size)
        for (i in 0 until size) {
            retVal[i] = `val`
        }
        return retVal
    }

    fun setBoolean(retVal: BooleanArray?, value: Boolean) {
        if (retVal?.indices != null) {
            for (i in retVal.indices) {
                retVal[i] = value
            }
        }
    }

    fun findMax(values: DoubleArray?): Double? {
        var d: Double? = null
        if (values != null) {
            for (v in values) {
                if (d == null || d < v) d = v
            }
        }
        return d
    }

    fun findMin(values: DoubleArray?): Double? {
        var d: Double? = null
        if (values != null) {
            for (v in values) {
                if (d == null || d > v) d = v
            }
        }
        return d
    }

    fun clone(src: ArrayList<Double?>?): MutableList<Double?>? {
        return if (src == null) {
            null
        } else ArrayList(src)
    }

    fun copy(origin: DoubleArray?, target: DoubleArray?, valueToAdd: Double) {
        if (origin?.indices != null) {
            for (i in origin.indices) {
                target?.set(i, origin[i] + valueToAdd)
            }
        }
    }
}